astrocut/cube_cut_from_footprint.py

```diff
@@ -0,0 +1,310 @@
+        List of paths to cutout files.
+    """
+    
+    # Convert to SkyCoord
+    if not isinstance(coordinates, SkyCoord):
+        coordinates = SkyCoord(coordinates, unit='deg')
+    if verbose:
+        print('Coordinates: ', coordinates)
+
+    # Parse cutout size
+    cutout_size = parse_size_input(cutout_size)
+    if verbose:
+        print('Cutout size: ', cutout_size)
+
+    # Get FFI footprints from the cloud
+    s3_uri = 's3://tesscut-ops-footprints/tess_ffi_footprint_cache.json' if product == 'SPOC' \
+        else 's3://tesscut-ops-footprints/tica_ffi_footprint_cache.json'
+    all_ffis = _get_s3_ffis(s3_uri=s3_uri, as_table=True, load_polys=True)
+    if verbose:
+        print(f'Found {len(all_ffis)} footprint files.')
+
+    # Filter FFIs by provided sectors
+    if sequences and product:
+        # Convert to list
+        if isinstance(sequences, int):
+            sequences = [sequences]
+        all_ffis = all_ffis[np.isin(all_ffis['sequence_number'], sequences)]
+
+        if len(all_ffis) == 0:
+            raise InvalidQueryError(f'No FFI cube files were found for sequences: \
+                                    {", ".join(str(s) for s in sequences)}')
+        
+        if verbose:
+            print(f'Filtered to {len(all_ffis)} footprints for sequences: {", ".join(str(s) for s in sequences)}')
+
+    # Get sector names and cube files that contain the cutout
+    cone_results = _ra_dec_crossmatch(all_ffis, coordinates, cutout_size, TESS_ARCSEC_PER_PX)
+    if not cone_results:
+        raise InvalidQueryError('The given coordinates were not found within the specified sector.')
+    seq_list = _create_sequence_list(cone_results, product)
+    cube_files_mapping = _get_cube_files_from_sequence_obs(seq_list)
+    if verbose:
+        print(f'Found {len(cube_files_mapping)} matching cube files.')
+    base_file_path = "s3://stpubdata/tess/public/mast/" if product == 'SPOC' \
+        else "s3://stpubdata/tess/public/mast/tica/"
+
+    # Make sure the output directory exists
+    if not os.path.exists(output_dir):
+        os.makedirs(output_dir)
+
+    # Executor function to generate cutouts from a cube file
+    def process_file(file):
+        try:
+            factory = CutoutFactory()
+            file_path = os.path.join(base_file_path, file['cube'])
+            output_path = os.path.join(output_dir, file['sectorName'])
+            cutout = factory.cube_cut(
+                file_path,
+                coordinates,
+                cutout_size=cutout_size,
+                product=product,
+                output_path=output_path,
+                threads='auto',
+                verbose=verbose
+            )
+            return cutout
+        except Exception as e:
+            warnings.warn(f'Unable to generate cutout from {file_path}: {e}', AstropyWarning)
+            return None
+        
+    # Generate cutout from each cube file
+    cutout_files = []
+    if threads == 'auto' or threads > 1:
```

Comment:
I am not sure threads here will help, and it may in fact hurt. `cube_cut` already uses threads for accessing the S3 data, so with this change, each cutout file would spawn that many threads. In my testing, there are diminishing returns after 8 threads. Since this could end up creating many times that number of threads, I expect we'd see thread contention here. 

If you set threads to 0, versus setting threads to "auto" or "8" what are the results here?

Reply:
Testing on my machine, I do see a performance improvement with a larger number of threads. It's more apparent when a sector isn't provided and more than 1 cutout is being generated. For example, these commands each generate 7 cutouts.

`cube_cut_from_footprint('130 30', cutout_size=50, output_dir='out_dir_2', threads=0)` --> 1 min, 23.4 sec
`cube_cut_from_footprint('130 30', cutout_size=50, output_dir='out_dir_2', threads=8)` --> 57.6 sec
`cube_cut_from_footprint('130 30', cutout_size=50, output_dir='out_dir_2', threads='auto')` --> 46.4 sec

Reply:
is this before or after you made the change to use the same threads variable to pass into the `cube_cut` function?

Reply:
Looks like I ran the test before, when `threads` was set to auto for `cube_cut`. When using the same `threads` variable, the call with `threads=0` takes a lot longer, which makes sense. I also see that there is less than a second difference between `threads=8` and `threads=auto`. 

Maybe it would be best to keep `threads` for `cube_cut` constant at some value, like 'auto' or 8? I think that using threads in `cube_cut_from_footprint` is still worthwhile for the performance improvement when making several cutouts at once, but performance does seem to stagnate after a certain point.